starknet-rpc-test/add_declare_transaction.rs

```diff
@@ -0,0 +1,154 @@
+    let (declare_tx, _, _) =
+        account.declare_contract("./contracts/Counter.sierra.json", "./contracts/Counter.casm.json");
+
+    let mut txs = madara.create_block_with_txs(vec![Transaction::Declaration(declare_tx)]).await?;
+
+    assert_eq!(txs.len(), 1);
+
+    let invoke_tx_result = txs.remove(0);
+    assert_matches!(
+        invoke_tx_result.err(),
+        Some(SendTransactionError::AccountError(starknet_accounts::AccountError::Provider(
+            ProviderError::StarknetError(StarknetErrorWithMessage {
+                code: MaybeUnknownErrorCode::Known(StarknetError::ValidationFailure),
+                message: _
+            })
+        )))
+    );
+
+    Ok(())
+}
+
+#[rstest]
+#[tokio::test]
+async fn fail_execution_step_with_no_storage_change(#[future] madara: MadaraClient) -> Result<(), anyhow::Error> {
+    let madara = madara.await;
+    let rpc = madara.get_starknet_client();
+
+    let account = create_account(rpc, SIGNER_PRIVATE, ARGENT_CONTRACT_ADDRESS, true);
+    let (declare_tx, expected_class_hash, _) =
+        account.declare_contract("./contracts/Counter.sierra.json", "./contracts/Counter.casm.json");
+
+    // draining account so the txn fails during execution
+    let balance =
+        read_erc20_balance(rpc, FieldElement::from_hex_be(FEE_TOKEN_ADDRESS).unwrap(), account.address()).await;
+    madara
+        .create_block_with_txs(vec![Transaction::Execution(account.transfer_tokens_u256(
+            FieldElement::from_hex_be("0x1234").unwrap(),
+            // subtractin 150k to keep some fees for the transfer
+            U256 { low: balance[0] - FieldElement::from_dec_str("150000").unwrap(), high: balance[1] },
+            None,
+        ))])
+        .await?;
+
+    // declaring contract
+    let mut txs = madara.create_block_with_txs(vec![Transaction::Declaration(declare_tx)]).await?;
+
+    assert_eq!(txs.len(), 1);
+    let declare_tx_result = txs.remove(0);
+    assert!(declare_tx_result.is_ok());
+
+    // transaction failed during execution, no change in storage
+    assert!(rpc.get_class(BlockId::Tag(BlockTag::Latest), expected_class_hash).await.is_err());
+
+    // doesn't get included in block
+    let included_txs = rpc.get_block_transaction_count(BlockId::Tag(BlockTag::Latest)).await?;
+    assert_eq!(included_txs, 0);
+
+    Ok(())
+}
+
+#[rstest]
+#[tokio::test]
+async fn works_with_storage_change(#[future] madara: MadaraClient) -> Result<(), anyhow::Error> {
+    let madara = madara.await;
+    let rpc = madara.get_starknet_client();
+
+    let account = create_account(rpc, SIGNER_PRIVATE, ARGENT_CONTRACT_ADDRESS, true);
+    let (declare_tx, expected_class_hash, _) =
+        account.declare_contract("./contracts/Counter.sierra.json", "./contracts/Counter.casm.json");
+
+    let mut txs = madara.create_block_with_txs(vec![Transaction::Declaration(declare_tx)]).await?;
+
+    assert_eq!(txs.len(), 1);
+    let declare_tx_result = txs.remove(0);
+    match declare_tx_result {
```

Comment:
`txs[0]`

Reply:
My (personal) opinion is that we don't really care about this kind of perf optimization for tests. The main focus should be the readability of the tests, so there is no mistake possible on what is happening and what it guarantees.
The less complexity, the less distraction the better. Technical complexity should be removed in order for the attention to focus on the test purpose.

So I would be totally ok to add an extra `.clone` if it allows for a more straightforward impl. 

What do you think?

Reply:
I understand your point about just focussing on readability for test cases, but what's the issue with `txs.remove(0)`?

Reply:
Remove gives you an `Option`. `[0]` is equivalent to `.remove(0).unwrap()` and it's ok because you know it is there.
Also `txs` does not have to be mutable`.

Imo it makes the intention clear by striping those unnecessary impl details.

But it's nothing truly important. If it causes other problems we can keep it

Reply:
Since I am using `remove` I need to make txs mutable. In this specific case, I could use `txs[0].as_ref` and clone the transaction hash when comparing. However, in some tests, I am comparing a lot of fields and will need to clone all of them to compare, so I feel having a `mut` with `remove` will be more readable. So I am keeping it for now, let me know though if you want me to change is particularly in some places.

Reply:
That's good to me